### DISCUSSION TITLE
Re-using the same layout.hbs without spamming my content files with {{> layout}}

### DISCUSSION THREAD

**Discussion by gruckionvit** [handlebars-lang/handlebars.js#1788] [Q&A]
Before filing issues, please check the following points first:

- [x] Please don't open issues for security issues. Instead, file a report at https://www.npmjs.com/advisories/report?package=handlebars
- [x] Have a look at https://github.com/wycats/handlebars.js/blob/master/CONTRIBUTING.md
- [x] Read the FAQ at https://github.com/wycats/handlebars.js/blob/master/FAQ.md
- [x] Use the jsfiddle-template at https://jsfiddle.net/4nbwjaqz/4/ to reproduce problems or bugs

This will probably help you to get a solution faster. 
For bugs, it would be great to have a PR with a failing test-case.

Hi, I would like to do something like what this guy has done.

https://github.com/handlebars-lang/handlebars.js/issues/1343

Specifically having a `layout.hbs` file for the layout (use case is sending emails). Then for each of the emails I want to trigger say `register.hbs` I then want to render register which is a section of the full HTML document nested inside the layouts file every time.

1. I use {{> layout}} on every single one of my content files, e.g. `register.hbs` `newsLetter.hbs` e.t.c.
2. I always renger `layout.hbs` and with a dynamic partial and then register every single one of my content files.

Both of these options don't seem right, I am not using express. When I Google my use case all I see is solutions with express or this `handlebars-layout` package which seems like massive over kill.

https://www.npmjs.com/package/handlebars-layout

Is there not a simple way of allowing me to wrap every template with the layout information? What options do I have?

```hbs
<!DOCTYPE html>
<html lang='en' prefix="og: http://ogp.me/ns#">

<head>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
	<meta http-equiv="X-UA-Compatible" content="IE=Edge">
	<meta name="robots" content="index, follow">
	<link rel="home" href="/">
	<meta name="viewport" content="initial-scale=1, width=device-width, maximum-scale=1, user-scalable=no">
	
	<title>{{title}}</title>

	<link type='text/css' rel='stylesheet' href='/css/materialize.min.css' />
	<link type='text/css' rel='stylesheet' href='/css/app.min.css' />
</head>

<body>
	{{> header }}

	{{{ content }}}

	{{> footer }}

	<script type='text/javascript' src='/js/jquery.min.js'></script>
	<script type='text/javascript' src='/js/materialize.min.js'></script>
	<script type='text/javascript' src='/js/primus.js'></script>
	<script type='text/javascript' src='/js/app.min.js'></script>
</body>

</html>
```

Example of a piece of content, there will be many different names templates. Any one of these templates can be rendered in the place where {{{ content }}} is.

```hbs
<div class='container'>
	<div class='row'>
		<div class='col s12 m7 l8'>
			<div class="card">
				<div class="card-content">
					<p>I am a very simple card. I am good at containing small bits of information. I am convenient because I require little
						markup to use effectively.</p>
				</div>
			</div>
		</div>

		<div class='col s12 m5 l4'>
			<div class="card">
				<div class="card-content">
					<p>I am a very simple card. I am good at containing small bits of information. I am convenient because I require little
						markup to use effectively.</p>
				</div>
			</div>
		</div>
	</div>
</div>
```

**Comment by gruckionvit** [comment]
Okay I found a solution. Please let me know if there is a better way of doing this.


We can effectively thing of the layout file as being the below, obviously we can do w.e we like in this file. But the main aim here is to get it to load the content from somewhere else.

```hbs
<html>
<body>
{{{ content }}}
</body>
</html>
```

For the partials (part of a html page not the HBS partial concept).

```hbs
<p>Hello {{ name }}</p>
``` 

Then use compile twice, once for the inner content where the content contains `name`.

```ts
export const getHtmlForTemplate = async (templateName: string, context: object): Promise<string> => {
  // get's the string content of the file by the templateName
  const partialSource = await getHtmlForPartial(templateName);

  try {
    const template = Handlebars.compile(partialSource);

    const partialTemplate = template(context);
    return getLayoutHtmlForTemplate(partialTemplate);
  } catch (error) {
    // Error parsing the template with the provided context
    throw new Error(error)
  }
}
```

Notice how the content being returned here is first parse through `getLayoutHtmlForTemplate(text: string)`

I then feed it through in much the way way.

```ts
export const getLayoutHtmlForTemplate = async (partialTemplate: string): Promise<string> => {
  // first get the layouts file content
  const layoutSource = await getHtmlForLayout();

  try {
    const template = Handlebars.compile(layoutSource)

    // process the layout template and place all of the content from the above function inside it. This will be rendered using the escape syntax `{{{ }}}` based on the `layout.hbs` file above.
    return template({ content: partialTemplate });
  } catch (error) {
    // Error parsing the template with the provided context
    throw new Error(error)
  }
}
```

Is there a better way than just double feeding? This seems fine, I pick up on the usage of `{{{ }}}` from the linked issue in my initial post.

**Comment by yespire** [comment]
I believe a new method assemblePartials() can make it easy, please check out my feature request:
https://github.com/handlebars-lang/handlebars.js/issues/1804
